icicle/include/icicle/program/program.h

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+#include <vector>
+#include <functional>
+#include "icicle/program/symbol.h"
+
+namespace icicle {
+
+  using InstructionType = uint32_t;
+
+  enum PreDefinedPrograms { AB_MINUS_C, EQ_X_AB_MINUS_C };
+
+  /**
+   * @brief A class that convert the function described by user into a program that can be executed
+   *
+   * This class recieves a Symbol instance that contains a DFG representing the required calculation.
+   * It generates a vector of instructions that represent the calculation.
+   * Each instruction has the following format.
+   * bits 7:0   - opcode according to enum OpCode
+   * bits 15:8  - operand 1 selector from the input vector
+   * bits 23:16 - operand 2 selector from the input vector
+   * bits 31:24 - result selector
+   */
+  template <typename S>
+  class Program
+  {
+  public:
+    // Generate a program based on a lambda function
+    Program(std::function<Symbol<S>(std::vector<Symbol<S>>&)> program_func, const int nof_inputs)
+    {
+      std::vector<Symbol<S>> program_inputs(nof_inputs);
+      set_as_inputs(program_inputs);
+      Symbol<S> result = program_func(program_inputs);
+      generate_program(result);
+    }
+
+    // Generate a program based on a PreDefinedPrograms
+    Program(PreDefinedPrograms pre_def)
+    {
+      switch (pre_def) {
```

Comment:
Is it sufficient to use only m_nof_imputs here to distinguish between cases?

Reply:
this are the predefined functions. The functionality is defined at the executors.
Only the number of inputs and outputs are missing